chore(repo/find): Exlude paths based on regex

BapRx opened the pull request:
Hi @hakoerber, I'm not a rust dev so there might be a better way to implement this (especially the case when no exclusion pattern is provided) but here's my PR to exclude paths from the `grm repos find local` discovery.

I'm open for suggestions to improve this :+1:

hakoerber commented:
Hey, very cool, thank you for your contribution!

### Documentation

It would be nice to advertise the new functionality a bit. Maybe add a short sentence in the "Generate your own configuration" section in `docs/src/local_configuration.md`. Maybe even with a small example.

### Testing

A test would be good. Take a look at `e2e_tests/test_repos_find.py`, especially the `test_repos_find()` test function. To me, there are three test cases that should be covered:
  * No regex is given => No change compared to current functionality (this one is already covered by the existing test cases).
  * A regex is given, and it matches one or more of the found repositories => All but the matched repositories need to be returned.
  * A regex is given, and it matches none of the found repositories => All repositories need to be returned.

So I'd do the following:

* Add a new parameter to `test_repos_find()` that has three values: `None`, `"match"`, `"nomatch"`.
  * `None` simply does nothing.
  * `match` adds `--exclude repo1` and checks that **only** repo2 is in the output
  * `nomatch` adds `--exclude foobar` and checks that both `repo1` and `repo2` are in the output (i.e. no change to the output check).

Honestly, the e2e test code is very ... special, and hard to get into. If you finish the rust side, I'm happy to do the changes to the tests.

---

> I'm open for suggestions to improve this

I'll take a closer look at the code itself and add comments directly if I find something that can be improved.

> I'm not a rust dev

Now you are ;) Feel free to add yourself to the `CONTRIBUTORS` file if you want.

hakoerber commented on src/lib.rs:
```diff
@@ -19,12 +19,23 @@ pub mod worktree;
-fn find_repos(root: &Path) -> Result<Option<(Vec<repo::Repo>, Vec<String>, bool)>, String> {
+fn find_repos(
+    root: &Path,
+    exclusion_pattern: &Option<String>,
```
Instead of taking a reference to an option, it's a bit more ergonomic to instead take an option that **contains** a reference. Instead of `&Option<T>`, you'd generally use `Option<&T>`.

And because we are handling strings here, instead of passing `&String`, you'd rather pass a slice (`&str`). That makes it easier for the caller to pass strings as-is (e.g. `find_repos(foo, Some("myregex"))` without further mangling. The caller never needs actual ownership of the `String`.

This also means that you can get rid of the `clone()` calls in line 31, as you are only passing string slices (i.e. references) around.

Note that this is completely irrelevant for the logic of the code in this case. It's just something good to know!

```suggestion
    exclusion_pattern: Option<&str>,
```

hakoerber commented on src/grm/main.rs:
```diff
@@ -199,7 +199,7 @@ fn main() {
                     };
 
-                    let (found_repos, warnings) = match find_in_tree(&path) {
+                    let (found_repos, warnings) = match find_in_tree(&path, &args.exclude) {
```
When changing `find_repos()` to take `Option<&str>`, you'll also have to update this. Now the question is, how do we transform `Option<String>` to `Option<&str>`?

Luckily, `String` implements `Deref<Target = str>`, so we can use a method already defined on `Option`:

```suggestion
                    let (found_repos, warnings) = match find_in_tree(&path, args.exclude.as_deref()) {
```

hakoerber commented on src/lib.rs:
```diff
@@ -19,12 +19,23 @@ pub mod worktree;
 
+    let exlusion_regex: regex::Regex =
+        regex::Regex::new(&exclusion_pattern.clone().unwrap_or(r"^$".to_string()))
+            .unwrap_or_else(|_| regex::Regex::new(r"^$").unwrap());
```
In case of an invalid regex, you'd silently discard the error and fall back to `^$`. I think it would be preferable to abort early with a proper error message, so the user can fix their regex.

You could do it like this:

`regex::Regex::new()` returns `Result<Regex, Error>`. If we get an error, we could immediately return this by using [the `?` operator](https://doc.rust-lang.org/book/ch09-02-recoverable-errors-with-result.html#a-shortcut-for-propagating-errors-the--operator). But `find_repos()` returns `Result<..., String>`. How can we convert this?

Take a look at [`Result::map_err()`](https://doc.rust-lang.org/std/result/enum.Result.html#method.map_err). It applies a function to the `Err` value of a `Result`, and leaves the `Ok` variant unmodified. Perfect for us.

So, you could change the code as follows:

```suggestion
            .map_err(|e| format!("invalid regex: {e}"))?;
*

hakoerber reviewed (commented):
Hey, I took a look at the code. The logic itself is sound, but there are few things that can make the code a bit more ergonomic. Take a look! If you have questions, let me know.

BapRx commented:
Hey @hakoerber, thank you very much for taking the time to explain this. It's very interesting and I tried to follow your recommendations.
I also added an example in the doc, linted the pytests and added a few conditions in the `test_repos_find()` function.

hakoerber commented on e2e_tests/test_repos_find.py:
```diff
@@ -101,9 +102,11 @@ def test_repos_find(configtype, default):
         args = ["repos", "find", "local", tmpdir]
         if not default:
             args += ["--format", configtype]
+            if exclude:
```
Any reason why you make `--exclude` depend on `default`?

`default` should actually be named `default_format` and just specify whether we *explicitly* pass `--format` or not. This should be independent of `--exclude`. So I think you can de-indent this if condition by one level.

This also means that you can drop the `and not default`s further down.

BapRx replied:
I understood the default flag as "default grm command without any other flag" and that's why I did it the same way as the format flag.
It's more clear now, I'm pushing an update to fix this.

hakoerber commented on e2e_tests/test_repos_find.py:
```diff
@@ -101,9 +102,11 @@ def test_repos_find(configtype, default):
         cmd = grm(args)
         assert cmd.returncode == 0
-        assert len(cmd.stderr) == 0
+        assert len(cmd.stderr) == (37 if exclude == "^.*/repo2$" and not default else 0)
```
I tried to not be too exact when matching error output. I think it's preferable to just check for some words that you *know* must be in the output:

```python
if exclude == "^.*/repo2$":
    assert "[skipped]" in cmd.stderr.lower()
    assert "repo2" in cmd.stderr.lower()
else:
    assert len(cmd.stderr) == 0
```

This makes the tests a bit more robust against small changes in the code. Otherwise, something innocent like adding full stops at the end of the error output would break the tests.

hakoerber commented on src/lib.rs:
```diff
@@ -19,12 +19,22 @@ pub mod worktree;
     let mut repo_in_root = false;
     let mut warnings = Vec::new();
 
+    let exlusion_regex: regex::Regex = regex::Regex::new(&exclusion_pattern.unwrap_or(r"^$"))
```
Clippy complains about this (this expression creates a reference which is immediately dereferenced by the compiler), and want to see this changed to the following:

```suggestion
    let exlusion_regex: regex::Regex = regex::Regex::new(exclusion_pattern.unwrap_or(r"^$"))
```

Short explanation: The `unwrap_or()` will already return a `&str` (as this is the content of the `Option`). There is no point in passing a reference of a `&str` to `Regex::new()`. This would be `&&str`. So in this case, the `&` is not needed.

hakoerber commented:
Absolutely awesome, thank you for this! Very cool that you also added e2e tests for this :+1: 

There are just a few small things to fix before this can be merged. I already added some inline comments.

One thing that cannot be done inline: `ruff` complains about unncecessary imports in `e2e_tests/docker-rest/flask/app.py`. But these imports (`github` and `gitlab`) are actually necessary for the flask testing API to function correctly (and they *have* to be at the end of the file, `isort` and `ruff` are wrong here). A simple `# noqa: E402,F401` at the end of those two imports should fix this.

---

All in all, as soon as `just check` runs succssfully, I am more than happy to merge this!

BapRx commented:
`just check` is happy :slightly_smiling_face: Thank you once again for the explanations, it's always very clear!

hakoerber commented:
Yes, looks good! Merged.

Thank you for your contribution :)

If you want me to add you to the `CONTRIBUTORS` file, let me know (also what I should put in there, like your github username or something else).

> Thank you once again for the explanations, it's always very clear!

My pleasure.
